### PR TITLE
Add simpleworker test

### PR DESCRIPTION
**Pull request by KimKiHyuk**
The issue is here #31 

It has a lot of points that should be refactored

if you think this test way is appropriate, I'm going to do some refactoring

### CODE REVIEW

**Comment by KimKiHyuk** [comment]
test failed check, WIP

**Comment by KimKiHyuk** [comment]
umm.. I'm a little bit confused 

CI server said 
```
System.Exception : JVM method execution failed: Nonstatic method 'collectToPython' failed for class '134' when called with no arguments
---- Microsoft.Spark.JvmException : org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 17.0 failed 1 times, most recent failure: Lost task 0.0 in stage 17.0 (TID 17, localhost, executor driver): org.apache.spark.SparkException: Python worker did not connect back in time
	at org.apache.spark.api.python.PythonWorkerFactory.createSimpleWorker(PythonWorkerFactory.scala:143)
	at org.apache.spark.api.python.PythonWorkerFactory.create(PythonWorkerFactory.scala:72)
	at org.apache.spark.SparkEnv.createPythonWorker(SparkEnv.scala:117)
	at org.apache.spark.api.python.BasePythonRunner.compute(PythonRunner.scala:86)
	at org.apache.spark.sql.execution.python.BatchEvalPythonExec.evaluate(BatchEvalPythonExec.scala:73)
	at org.apache.spark.sql.execution.python.EvalPythonExec$$anonfun$doExecute$1.apply(EvalPythonExec.scala:131)
	at org.apache.spark.sql.execution.python.EvalPythonExec$$anonfun$doExecute$1.apply(EvalPythonExec.scala:93)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitions$1$$anonfun$apply$23.apply(RDD.scala:800)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitions$1$$anonfun$apply$23.apply(RDD.scala:800)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:87)
	at org.apache.spark.scheduler.Task.run(Task.scala:109)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:345)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.net.SocketTimeoutException: Accept timed out
	at java.net.DualStackPlainSocketImpl.waitForNewConnection(Native Method)
	at java.net.DualStackPlainSocketImpl.socketAccept(DualStackPlainSocketImpl.java:135)
	at java.net.AbstractPlainSocketImpl.accept(AbstractPlainSocketImpl.java:409)
	at java.net.PlainSocketImpl.accept(PlainSocketImpl.java:199)
	at java.net.ServerSocket.implAccept(ServerSocket.java:560)
	at java.net.ServerSocket.accept(ServerSocket.java:528)
	at org.apache.spark.api.python.PythonWorkerFactory.createSimpleWorker(PythonWorkerFactory.scala:138)
	... 23 more

```

but it's ok with my machine.

image


could you give me a clue?

**Comment by KimKiHyuk** [comment]
@imback82 Can i ask you to review it?

File: src/csharp/Microsoft.Spark.Worker.UnitTest/SimpleWorkerTests.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+using Xunit;
+
+namespace Microsoft.Spark.Worker.UnitTest
+{
+    public class SimpleWorkerTests
+    {
+        [Fact]
+        public void TestsSimpleWorkerTaskRunners()
+        {
+            var typedVersion = new Version(Versions.V2_4_0);
```
Comment (imback82) [inline src/csharp/Microsoft.Spark.Worker.UnitTest/SimpleWorkerTests.cs:15]: Test against different versions?

File: src/csharp/Microsoft.Spark.Worker.UnitTest/SimpleWorkerTests.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+            ISocketWrapper serverListen = SocketFactory.CreateSocket();
+            var ipEndpoint = (IPEndPoint)serverListen.LocalEndPoint;
+            int port = ipEndpoint.Port;
+
+            serverListen.Listen();
+
+            PayloadWriter payloadWriter = new PayloadWriterFactory().Create();
+            Task clientTask = Task.Run(() => simpleWorker.Run(port));
+
+            using (ISocketWrapper serverSocket = serverListen.Accept())
```
Comment (imback82) [inline src/csharp/Microsoft.Spark.Worker.UnitTest/SimpleWorkerTests.cs:26]: Can we merge with TaskRunnerTests which has the same test?

Comment (KimKiHyuk) [reply src/csharp/Microsoft.Spark.Worker.UnitTest/SimpleWorkerTests.cs:28]: sure! I'm working on it now

Comment (KimKiHyuk) [reply src/csharp/Microsoft.Spark.Worker.UnitTest/SimpleWorkerTests.cs]: resolved at 4ececb7

File: src/csharp/Microsoft.Spark.Worker.UnitTest/SimpleWorkerTests.cs (hunk 1 above)
Comment (KimKiHyuk) [reply src/csharp/Microsoft.Spark.Worker.UnitTest/SimpleWorkerTests.cs]: resolved at https://github.com/dotnet/spark/pull/644/commits/bcff57b554dc911cc238f5c344b86b6c9efbf19e

File: src/csharp/Microsoft.Spark.Worker/SimpleWorker.cs [hunk 3]
```diff
@@ -22,18 +22,24 @@ internal SimpleWorker(Version version)
...
+        }
+
+        internal void Run(int port)
         {
             try
             {
-                int port = Utils.SettingUtils.GetWorkerFactoryPort(_version);
                 string secret = Utils.SettingUtils.GetWorkerFactorySecret(_version);
 
+                ISocketWrapper clientSocket = SocketFactory.CreateSocket();
```
[the remark this answers is not in the document]
Comment (KimKiHyuk) [reply src/csharp/Microsoft.Spark.Worker/SimpleWorker.cs]: I'm with you, it has been resolved at a7c7999

**Comment by KimKiHyuk** [comment]
@imback82

https://github.com/dotnet/spark/pull/656

CI works properly when removing 2.4.6 from testcase
here is the result of CI failed
https://dev.azure.com/dnceng/public/_build/results?buildId=801608&view=logs&jobId=be81569a-8125-5d8e-166f-c31637c865ad&j=be81569a-8125-5d8e-166f-c31637c865ad&t=0211fc11-88de-58c5-af5f-02daa74d8dcb

CI has been passed after https://github.com/dotnet/spark/pull/644/commits/aa4df4cd3920982a3e400d67e084ba2eca09300c which removing 2.4.6 from test cases.

If #656 will be merged, I'm going to re-add 2.4.6

**Comment by KimKiHyuk** [comment]
#656 seems unchanged, 

all jobs done

File: src/csharp/Microsoft.Spark/Versions.cs [hunk 4]
```diff
@@ -12,6 +12,8 @@ internal static class Versions
         internal const string V2_3_3 = "2.3.3";
         internal const string V2_4_0 = "2.4.0";
         internal const string V2_4_2 = "2.4.2";
+        internal const string V2_4_4 = "2.4.4";
+        internal const string V2_4_5 = "2.4.5";
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Versions.cs:16]: Why are these added?

Comment (KimKiHyuk) [reply src/csharp/Microsoft.Spark/Versions.cs]: If i remerber right,  E2E test was failed on stage V2_4_4 and V_2_4_5 

I've pushed new commit which removed V_2_4._4 V_2_4_5 

let's see CI result

Comment (KimKiHyuk) [reply src/csharp/Microsoft.Spark/Versions.cs]: I got it wrong,  it looks good.

File: src/csharp/Microsoft.Spark.Worker.UnitTest/TaskRunnerTests.cs [hunk 5]
```diff
@@ -26,6 +26,13 @@ public void TestTaskRunner()
             var taskRunner = new TaskRunner(0, clientSocket, false, payloadWriter.Version);
             Task clientTask = Task.Run(() => taskRunner.Run());
 
+            TestTaskRunnerReadWrite(serverListener, payloadWriter);
+
+            Assert.True(clientTask.Wait(5000));
+        }
+
+        static internal void TestTaskRunnerReadWrite(ISocketWrapper serverListener, PayloadWriter payloadWriter)
```
Comment (suhsteve) [inline src/csharp/Microsoft.Spark.Worker.UnitTest/TaskRunnerTests.cs:34]: format line.  Over 100 chars.  Also `internal static` vs `static internal`

File: src/csharp/Microsoft.Spark.Worker.UnitTest/SimpleWorkerTests.cs [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+{
+    [Collection("Spark Unit Tests")]
+    public class SimpleWorkerTests
+    {
+        [Theory]
+        [MemberData(nameof(TestData.VersionData), MemberType = typeof(TestData))]
+        public void TestsSimpleWorkerTaskRunners(string version)
+        {
+            var typedVersion = new Version(version);
+            var simpleWorker = new SimpleWorker(typedVersion);
```
Comment (suhsteve) [inline src/csharp/Microsoft.Spark.Worker.UnitTest/SimpleWorkerTests.cs:18]: move this closer to where it's used.

File: src/csharp/Microsoft.Spark.Worker.UnitTest/SimpleWorkerTests.cs [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+            using ISocketWrapper serverListener = SocketFactory.CreateSocket();
+            var ipEndpoint = (IPEndPoint)serverListener.LocalEndPoint;
+            int port = ipEndpoint.Port;
+
+            serverListener.Listen();
+
+            PayloadWriter payloadWriter = new PayloadWriterFactory().Create(typedVersion);
```
Comment (suhsteve) [inline src/csharp/Microsoft.Spark.Worker.UnitTest/SimpleWorkerTests.cs:25]: move this closer to where it's used.

File: src/csharp/Microsoft.Spark.Worker.UnitTest/SimpleWorkerTests.cs [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+            Task clientTask = Task.Run(() => simpleWorker.Run(port));
```
Comment (suhsteve) [inline src/csharp/Microsoft.Spark.Worker.UnitTest/SimpleWorkerTests.cs:26]: you can use `ipEndpoint.port` directly.

File: src/csharp/Microsoft.Spark.Worker.UnitTest/TaskRunnerTests.cs (hunk 5 above)
Comment (KimKiHyuk) [reply src/csharp/Microsoft.Spark.Worker.UnitTest/TaskRunnerTests.cs:34]: internal static looks good to me

File: src/csharp/Microsoft.Spark.Worker/SimpleWorker.cs [hunk 9]
```diff
@@ -22,15 +22,21 @@ internal SimpleWorker(Version version)
...
+
+        internal void Run(int port)
         {
             try
             {
-                int port = Utils.SettingUtils.GetWorkerFactoryPort(_version);
                 string secret = Utils.SettingUtils.GetWorkerFactorySecret(_version);
 
+                ISocketWrapper socket = SocketFactory.CreateSocket();
+
```
[the remark this answers is not in the document]
Comment (KimKiHyuk) [reply src/csharp/Microsoft.Spark.Worker/SimpleWorker.cs:37]: Nope. going to revert it

File: src/csharp/Microsoft.Spark.Worker.UnitTest/SimpleWorkerTests.cs (hunk 6 above)
Comment (KimKiHyuk) [reply src/csharp/Microsoft.Spark.Worker.UnitTest/SimpleWorkerTests.cs]: resolved at 
3b8c222

File: src/csharp/Microsoft.Spark.Worker.UnitTest/SimpleWorkerTests.cs (hunk 7 above)
Comment (KimKiHyuk) [reply src/csharp/Microsoft.Spark.Worker.UnitTest/SimpleWorkerTests.cs]: resolved at 
3b8c222

File: src/csharp/Microsoft.Spark.Worker.UnitTest/SimpleWorkerTests.cs (hunk 8 above)
Comment (KimKiHyuk) [reply src/csharp/Microsoft.Spark.Worker.UnitTest/SimpleWorkerTests.cs]: resolved at 
2ec6d77

File: src/csharp/Microsoft.Spark.Worker.UnitTest/TaskRunnerTests.cs (hunk 5 above)
Comment (KimKiHyuk) [reply src/csharp/Microsoft.Spark.Worker.UnitTest/TaskRunnerTests.cs]: resolved at 
7e715ce

File: src/csharp/Microsoft.Spark.Worker/SimpleWorker.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (KimKiHyuk) [reply src/csharp/Microsoft.Spark.Worker/SimpleWorker.cs]: resolved at 
e12ccb2

File: src/csharp/Microsoft.Spark.Worker/SimpleWorker.cs [hunk 10]
```diff
@@ -22,15 +22,20 @@ internal SimpleWorker(Version version)
...
+            Run(port);
+        }
+
+        internal void Run(int port)
         {
             try
             {
-                int port = Utils.SettingUtils.GetWorkerFactoryPort(_version);
                 string secret = Utils.SettingUtils.GetWorkerFactorySecret(_version);
+                ISocketWrapper socket = SocketFactory.CreateSocket();
```
Comment (suhsteve) [inline src/csharp/Microsoft.Spark.Worker/SimpleWorker.cs:35]: can you move this back above `socket.Connect(IPAddress.Loopback, port, secret);` ?

Comment (KimKiHyuk) [reply src/csharp/Microsoft.Spark.Worker/SimpleWorker.cs]: resolved at 89cc504

**Review by suhsteve — APPROVED** [review]
LGTM.  Thanks @KimKiHyuk !

**Review by imback82 — APPROVED** [review]
LGTM, thanks @KimKiHyuk!
